File: lib/Utils/PollingTask.cs
// Copyright (c) 2020 ayuma_x. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.Threading;
using System.Threading.Tasks;

namespace ObjectDeliverer.Utils
{
    class PollingTask : IAsyncDisposable
    {
        private ValueTask pollingTask = default(ValueTask);

        private CancellationTokenSource? canceler;

        PollingTask(Func<ValueTask<bool>> action)
        {
            this.canceler = new CancellationTokenSource();

            this.pollingTask = this.RunAsync(action);
        }

        ValueTask DisposeAsync()
        {
            if (this.pollingTask == null)return default(ValueTask);

            this.canceler?.Cancel();
            return this.pollingTask;
        }

        private async ValueTask RunAsync(Func<ValueTask<bool>> action)
        {
            while (this.canceler?.IsCancellationRequested == false)
            {
                if (await action.Invoke() == false)
                {
                    break;
                }

                await Task.Delay(1);
            }
        }
    }
}

File: lib/Utils/polling_task.dart
// Copyright (c) 2020 ayuma_x. All rights reserved.
// Licensed under the BSD license. See LICENSE file in the project root for full license information.

import 'package:async/async.dart';

class PollingTask {
  PollingTask.fromAction(Future<bool> Function() action) {
    _isCancel = false;

    _pollingTask = CancelableOperation<void>.fromFuture(runAsync(action),
        onCancel: _onCancel);
  }

  CancelableOperation<void> _pollingTask;
  bool _isCancel = false;

  Future<void> disposeAsync() async {
    if (_pollingTask == null) {
      return;
    }

    _pollingTask.cancel();

    return _pollingTask.value;
  }

  Future<void> runAsync(Future<bool> Function() action) async {
    while (_isCancel == false) {
      if (await action() == false) {
        break;
      }

      await Future<void>.delayed(const Duration(milliseconds: 1));
    }
  }

  void _onCancel() {
    _isCancel = true;
  }
}
